plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java

```diff
@@ -2885,14 +2885,15 @@ else if (isFullAcidTable(table.getParameters())) {
                 .map(Column::getName)
                 .forEach(partitioningColumns::add);
 
+        ImmutableList<HiveType> bucketingColumnTypes = hiveBucketHandle.get().getColumns().stream()
+                .map(HiveColumnHandle::getHiveType)
+                .collect(toImmutableList());
         HivePartitioningHandle partitioningHandle = new HivePartitioningHandle(
                 hiveBucketHandle.get().getBucketingVersion(),
                 hiveBucketHandle.get().getTableBucketCount(),
-                hiveBucketHandle.get().getColumns().stream()
-                        .map(HiveColumnHandle::getHiveType)
-                        .collect(toImmutableList()),
+                bucketingColumnTypes,
                 OptionalInt.of(hiveBucketHandle.get().getTableBucketCount()),
-                !partitionColumns.isEmpty() && isParallelPartitionedBucketedWrites(session));
+                !partitionColumns.isEmpty() && !bucketingColumnTypes.isEmpty() && isParallelPartitionedBucketedWrites(session));
```

Comment:
I don't understand why `&& !bucketingColumnTypes.isEmpty()` is added here.

cc @raunaqmorarka @sopel39

Reply:
In my understanding because we started to rely more on `HivePartitioningHandle' and its `isSingleNode` method and I wanted to `isUsePartitionedBucketing` to return value that is consistent with `isSingleNode` - if `isSingleNode` returns true than `isUsePartitionedBucketing` should return false

Reply:
it seems that `isUsePartitionedBucketing` is now false for partitioned, unbucketed, non-transactional tables, while it used to be true.

Reply:
unbacketed means `hiveBucketHandle.isEmpty()` So it looks we should bail out of the method earlier and never get here.

Reply:
> unbacketed means hiveBucketHandle.isEmpty() So it looks we should bail out of the method earlier and never get here.

Yes. I think we should hit `if (hiveBucketHandle.isEmpty()) {` earlier in the code, so I'm not sure why this check is needed

Reply:
@sopel39  this check is for the case when there is  an implicit bucketing with one bucket without bucketing function - this happens for transactional table

Reply:
> @sopel39 this check is for the case when there is an implicit bucketing with one bucket without bucketing function - this happens for transactional table

So it seems we would like to distribute writes across partitions in this case. Otherwise, data will be written with single writer only in entire cluster

Reply:
No - we need to generate single bucket file then. So we need single writer. (IIUC)

Reply:
> No - we need to generate single bucket file then. So we need single writer. (IIUC)

Single bucket per partition, right? So `N` files, where `N` is partition count

Reply:
I probably don't understand something but I just tested and partitioned transactional table(so 1 implicit bucket) seems to work fine, different partitions are created and each of them have 1 bucket

Reply:
I can add test for that if necessary

Reply:
> I probably don't understand something but I just tested and partitioned transactional table(so 1 implicit bucket) seems to work fine, different partitions are created and each of them have 1 bucket

Writes will be correct, but with the change here only one node and one thread (in entire cluster) will be writing data. The code you changed distributes writes between worker nodes, so we can avoid single writer in entire cluster.

Reply:
@homar it sounds like you tested **transactional** partitioned tables (unbucketed; aka with implicit 1 bucket)
The concern is about INSERT into **non-transactional** partitioned, unbucketed table.

Reply:
> @homar it sounds like you tested transactional partitioned tables (unbucketed; aka with implicit 1 bucket)
The concern is about INSERT into non-transactional partitioned, unbucketed table.

I think we want to redistribute writes even for transactional, partitioned and bucketed (implicit 1 bucket).

Reply:
> hiveBucketHandle.isEmpty()

unbucketed, non-transactional should not reach this code beacause of `if (hiveBucketHandle.isEmpty()) {` earlier in the code. But unfortunately I am afraid that @sopel39 comments regarding decreasing number of writers is still a valid concern.

Reply:
@sopel39 again I probably miss something but even with my changes for transactional, unbucketed(so 1 implicit bucket) and partitioned table when I try to make an insert that creates 100 partitions here `https://github.com/trinodb/trino/blob/ee2ef32e6f09515a888a016adc1cc6ccd32cbae4/plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePageSink.java#L354` I get 100 writers. Maybe you mentioned different writers - in such a case please point me to the part of particular part of the code